test: add tests for IEEE 754-2019 compliance

anandkaranubc opened the pull request:
Progresses #365

## Description

> What is the purpose of this pull request?

This pull request:

- Adds all the tests for testing special cases of `log*`

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   progresses #365

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

anandkaranubc committed:
test: add tests for IEEE 754-2019 compliance

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc committed:
refactor: test messages

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

anandkaranubc commented:
<img width="816" alt="image" src="https://github.com/user-attachments/assets/e43ee498-e7b7-4ee3-bbf0-dfe1633af1ea" />

All checked!

Planeshifter commented:
@anandkaranubc Per failing CI checks, looks like our `log10` might not behave as expected for `-0`. Would you mind taking a look, please?

anandkaranubc commented:
> Haha, was in the middle of writing the reason :)

Oops, there is a bug. `log2` and `log10` don't return `NINF` when `x = -0.0`. But it works in C.

Mainly because of no type casting on `hx` in JS.

https://github.com/stdlib-js/stdlib/blob/79667271d21f6ca4242a66f2bce5d884ee31fe93/lib/node_modules/%40stdlib/math/base/special/log2/lib/main.js#L118-L126

`hx` still remains a "huge" positive number here and isn't less than `HIGH_MIN_NORMAL_EXP`

But in C,

Typecasting it back again in `int32_t` makes it negative again. 

https://github.com/stdlib-js/stdlib/blob/79667271d21f6ca4242a66f2bce5d884ee31fe93/lib/node_modules/%40stdlib/math/base/special/log2/src/main.c#L71-L79

cc: @Planeshifter @kgryte

anandkaranubc committed:
Merge branch 'develop' into ieee/log

Signed-off-by: Karan Anand <119553199+anandkaranubc@users.noreply.github.com>

anandkaranubc committed:
test: add tests for IEEE 754-2019 compliance

PR-URL: https://github.com/stdlib-js/stdlib/pull/6599
Ref: https://github.com/stdlib-js/stdlib/issues/365

Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
docs: update list of contributors

PR-URL: https://github.com/stdlib-js/stdlib/pull/6590

Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>
Signed-off-by: stdlib-bot <82920195+stdlib-bot@users.noreply.github.com>

anandkaranubc committed:
test: add tests for IEEE 754-2019 compliance

PR-URL: https://github.com/stdlib-js/stdlib/pull/6595
Ref: https://github.com/stdlib-js/stdlib/issues/365

Reviewed-by: Philipp Burckhardt <pburckhardt@outlook.com>

stdlib-bot committed:
docs: update REPL namespace documentation

PR-URL: https://github.com/stdlib-js/stdlib/pull/6591
Reviewed-by: Athan Reines <kgryte@gmail.com>

stdlib-bot committed:
docs: update related packages sections

PR-URL: https://github.com/stdlib-js/stdlib/pull/6617
Reviewed-by: Athan Reines <kgryte@gmail.com>

hrshya committed:
docs: replace manual `for` loop in examples

PR-URL: https://github.com/stdlib-js/stdlib/pull/6612
Reviewed-by: Athan Reines <kgryte@gmail.com>

hrshya committed:
docs: replace manual `for` loop in examples

PR-URL: https://github.com/stdlib-js/stdlib/pull/6611
Reviewed-by: Athan Reines <kgryte@gmail.com>

hrshya committed:
docs: replace manual `for` loop in examples

PR-URL: https://github.com/stdlib-js/stdlib/pull/6607
Reviewed-by: Athan Reines <kgryte@gmail.com>

hrshya committed:
docs: replace manual `for` loop in examples

PR-URL: https://github.com/stdlib-js/stdlib/pull/6608
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: Athan Reines <kgryte@gmail.com>

hrshya committed:
docs: replace manual `for` loop in examples

PR-URL: https://github.com/stdlib-js/stdlib/pull/6587
Reviewed-by: Athan Reines <kgryte@gmail.com>

mhmohona committed:
chore: fix EditorConfig lint errors

PR-URL: https://github.com/stdlib-js/stdlib/pull/6596
Closes: https://github.com/stdlib-js/stdlib/issues/6588
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: Mahfuza Humayra Mohona <mhmohona@gmail.com>
Signed-off-by: Athan Reines <kgryte@gmail.com>

anandkaranubc committed:
test: add tests for IEEE 754-2019 compliance

PR-URL: https://github.com/stdlib-js/stdlib/pull/6597
Ref: https://github.com/stdlib-js/stdlib/issues/365
Reviewed-by: Athan Reines <kgryte@gmail.com>
Signed-off-by: Karan Anand <119553199+anandkaranubc@users.noreply.github.com>

headlessNode committed:
feat: add `array/base/reshape`

PR-URL: https://github.com/stdlib-js/stdlib/pull/5639
Closes: https://github.com/stdlib-js/metr-issue-tracker/issues/20
Co-authored-by: Athan Reines <kgryte@gmail.com>
Reviewed-by: Athan Reines <kgryte@gmail.com> 
Signed-off-by: Muhammad Haris <harriskhan047@outlook.com>
Signed-off-by: Athan Reines <kgryte@gmail.com>

kgryte commented:
@anandkaranubc For L22, if the condition assumes `ihx`, then we can modify to

```javascript
if ( (hx|0) < HIGH_MIN_NORMAL_EXP ) { 
 	// Case: x < 2**-1022 
 	if ( ( (hx&ABS_MASK) | lx ) === 0 ) { 
 		return NINF; 
 	} 
```

anandkaranubc commented:
@kgryte Thanks for the reply!

Can I do `hx = WORDS[ 0 ] | 0;`  instead?

anandkaranubc committed:
fix: use asm type annotation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented:
@anandkaranubc Actually, are we sure the issue is with `hx` here? Given 

https://github.com/stdlib-js/stdlib/blob/79667271d21f6ca4242a66f2bce5d884ee31fe93/lib/node_modules/%40stdlib/math/base/special/log2/lib/main.js#L115

`x` will be positive once going to `toWords`. Hence, the sign-bit of the high word will be zero. In which case, `hx` and `hx|0` should be the same value.

anandkaranubc commented:
/stdlib merge

kgryte commented:
@anandkaranubc Ah, I see. The special case is `-0`. In which case, yes, you are right. Based on the code, it should be reasonable to do

```
hx = WORDS[ 0 ] | 0; // asm type annotation
```

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ieee/log

anandkaranubc commented on lib/node_modules/@stdlib/math/base/special/log10/lib/main.js:
```diff
@@ -132,7 +132,7 @@ function log10( x ) {
 		// Subnormal number, scale up x:
 		x *= TWO54;
-		hx = getHighWord( x );
+		hx = getHighWord( x ) | 0; // asm type annotation
```
I also updated this line accordingly to be consistent with upstream (and C) implementation.

[Reference](https://github.com/freebsd/freebsd-src/blob/89e0e3814e8ca205db7523f6ce6c63cebe4bfb0d/lib/msun/src/e_log10.c#L51-L52)

anandkaranubc replied:
Our C implementation:

https://github.com/stdlib-js/stdlib/blob/864fc4ffe65a2f6455f92ac10947d9921e23c355/lib/node_modules/%40stdlib/math/base/special/log10/src/main.c#L104-L106

anandkaranubc commented:
The failing test is unrelated to this PR.

kgryte commented:
@anandkaranubc I'll fix that up on `develop` now.

kgryte commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into ieee/log

kgryte reviewed (approved):
LGTM
